<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Matthew Hung">
<meta name="dcterms.date" content="2025-01-22">
<meta name="description" content="Team10 MetaSix">

<title>3&nbsp; Risk Classification with ssGSEA – CRUK 2025 Hackathon</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./xenium2.html" rel="next">
<link href="./metastatic_signature2.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d1cbdf98a2738ec369a65716124baecb.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./ssgsea2.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Risk Classification with ssGSEA</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">CRUK 2025 Hackathon</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/FrancisCrickInstitute/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title"><strong>Introduction</strong></span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./metastatic_signature2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Generating Prognostic Signature</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ssgsea2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Risk Classification with ssGSEA</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./xenium2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Validation in Xenium</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./validation2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Validation of Xenium Genes</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#actions" id="toc-actions" class="nav-link active" data-scroll-target="#actions"><span class="header-section-number">3.1</span> Actions</a></li>
  <li><a href="#format-data-for-ssgsea" id="toc-format-data-for-ssgsea" class="nav-link" data-scroll-target="#format-data-for-ssgsea"><span class="header-section-number">3.2</span> Format data for ssGSEA</a></li>
  <li><a href="#perform-ssgsea" id="toc-perform-ssgsea" class="nav-link" data-scroll-target="#perform-ssgsea"><span class="header-section-number">3.3</span> Perform ssGSEA</a></li>
  <li><a href="#combine-ssgsea-with-survival" id="toc-combine-ssgsea-with-survival" class="nav-link" data-scroll-target="#combine-ssgsea-with-survival"><span class="header-section-number">3.4</span> Combine ssGSEA with survival</a></li>
  <li><a href="#lusc" id="toc-lusc" class="nav-link" data-scroll-target="#lusc"><span class="header-section-number">3.5</span> LUSC</a></li>
  <li><a href="#luad" id="toc-luad" class="nav-link" data-scroll-target="#luad"><span class="header-section-number">3.6</span> LUAD</a></li>
  <li><a href="#save-output" id="toc-save-output" class="nav-link" data-scroll-target="#save-output"><span class="header-section-number">3.7</span> Save Output</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.dev/FrancisCrickInstitute/blob/main/ssgsea2.ipynb" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Risk Classification with ssGSEA</span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Matthew Hung </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">2025-01-22</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="actions" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="actions"><span class="header-section-number">3.1</span> Actions</h2>
<p>We will now use the gene signatures to classify patients into low/high groups. This is done by:</p>
<ul>
<li>ssGSEA of gene sigature</li>
<li>splitting the score into 200 intervals and selecting the optimal cox-regression p value cutoff</li>
</ul>
</section>
<section id="format-data-for-ssgsea" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="format-data-for-ssgsea"><span class="header-section-number">3.2</span> Format data for ssGSEA</h2>
<div id="cell-3" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="1">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># processed on NEMO so far</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">library</span>(glmnet)</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">library</span>(survival)</span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="fu">setwd</span>(<span class="st">"/camp/home/hungm/scratch/hungm/matthew/MH_Hackathon/2025_CRUK/data/TCGA2/0_raw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors
Loading required package: MatrixGenerics

Loading required package: matrixStats


Attaching package: ‘matrixStats’


The following object is masked from ‘package:dplyr’:

    count



Attaching package: ‘MatrixGenerics’


The following objects are masked from ‘package:matrixStats’:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars


Loading required package: GenomicRanges

Loading required package: stats4

Loading required package: BiocGenerics


Attaching package: ‘BiocGenerics’


The following objects are masked from ‘package:lubridate’:

    intersect, setdiff, union


The following objects are masked from ‘package:dplyr’:

    combine, intersect, setdiff, union


The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs


The following objects are masked from ‘package:base’:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min


Loading required package: S4Vectors


Attaching package: ‘S4Vectors’


The following objects are masked from ‘package:lubridate’:

    second, second&lt;-


The following objects are masked from ‘package:dplyr’:

    first, rename


The following object is masked from ‘package:tidyr’:

    expand


The following object is masked from ‘package:utils’:

    findMatches


The following objects are masked from ‘package:base’:

    expand.grid, I, unname


Loading required package: IRanges


Attaching package: ‘IRanges’


The following object is masked from ‘package:lubridate’:

    %within%


The following objects are masked from ‘package:dplyr’:

    collapse, desc, slice


The following object is masked from ‘package:purrr’:

    reduce


Loading required package: GenomeInfoDb

Loading required package: Biobase

Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.



Attaching package: ‘Biobase’


The following object is masked from ‘package:MatrixGenerics’:

    rowMedians


The following objects are masked from ‘package:matrixStats’:

    anyMissing, rowMedians


Loading required package: Matrix


Attaching package: ‘Matrix’


The following object is masked from ‘package:S4Vectors’:

    expand


The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack


Loaded glmnet 4.1-8
</code></pre>
</div>
</div>
<div id="cell-4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="2">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">source</span>(<span class="st">"/camp/home/hungm/working/Matthew/library/R/functions/coxph/coxph.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-5" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># read gene expression</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>gex <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../1_processed/TCGA_lung_gex.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="fu">colnames</span>(gex) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="st">"-"</span>, <span class="fu">colnames</span>(gex))</span>
<span id="cb4-4"><a href="#cb4-4"></a>gex[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 5 × 5</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">TCGA-05-4244</th>
<th data-quarto-table-cell-role="th" scope="col">TCGA-05-4249</th>
<th data-quarto-table-cell-role="th" scope="col">TCGA-05-4250</th>
<th data-quarto-table-cell-role="th" scope="col">TCGA-05-4382</th>
<th data-quarto-table-cell-role="th" scope="col">TCGA-05-4384</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">A1BG</td>
<td>26.0302</td>
<td>120.1350</td>
<td>50.8597</td>
<td>145.904</td>
<td>127.3670</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">A1CF</td>
<td>0.0000</td>
<td>0.3220</td>
<td>0.0000</td>
<td>0.000</td>
<td>0.0000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">A2BP1</td>
<td>1.7454</td>
<td>1.6098</td>
<td>0.0000</td>
<td>0.000</td>
<td>0.0000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">A2LD1</td>
<td>135.5020</td>
<td>89.0629</td>
<td>151.1330</td>
<td>112.069</td>
<td>87.5748</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">A2M</td>
<td>9844.7900</td>
<td>25712.7000</td>
<td>16943.6000</td>
<td>23326.300</td>
<td>48314.6000</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>write.gct <span class="ot">&lt;-</span> <span class="cf">function</span>(df, outdir){</span>
<span id="cb5-2"><a href="#cb5-2"></a>    ncol <span class="ot">&lt;-</span> <span class="fu">ncol</span>(df)</span>
<span id="cb5-3"><a href="#cb5-3"></a>    nrow <span class="ot">&lt;-</span> <span class="fu">nrow</span>(df)</span>
<span id="cb5-4"><a href="#cb5-4"></a>    annot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">NAME =</span> <span class="fu">rownames</span>(df), <span class="at">Description =</span> <span class="fu">rownames</span>(df))</span>
<span id="cb5-5"><a href="#cb5-5"></a>    df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(annot, df)</span>
<span id="cb5-6"><a href="#cb5-6"></a>    header <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="st">""</span>, <span class="at">ncol =</span> <span class="fu">ncol</span>(df), <span class="at">nrow =</span> <span class="dv">2</span>))</span>
<span id="cb5-7"><a href="#cb5-7"></a>    <span class="fu">colnames</span>(header) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(df)</span>
<span id="cb5-8"><a href="#cb5-8"></a>    header[<span class="dv">3</span>,] <span class="ot">&lt;-</span> <span class="fu">colnames</span>(df)</span>
<span id="cb5-9"><a href="#cb5-9"></a>    header[<span class="dv">1</span>,<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"#1.2"</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>    header[<span class="dv">2</span>,<span class="dv">1</span>] <span class="ot">&lt;-</span> nrow</span>
<span id="cb5-11"><a href="#cb5-11"></a>    header[<span class="dv">2</span>,<span class="dv">2</span>] <span class="ot">&lt;-</span> ncol</span>
<span id="cb5-12"><a href="#cb5-12"></a>    df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(header, df)</span>
<span id="cb5-13"><a href="#cb5-13"></a>    <span class="fu">write.table</span>(df, <span class="at">file =</span> outdir, <span class="at">quote =</span> F, <span class="at">row.names =</span> F, <span class="at">col.names =</span> F, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb5-14"><a href="#cb5-14"></a>}</span>
<span id="cb5-15"><a href="#cb5-15"></a></span>
<span id="cb5-16"><a href="#cb5-16"></a><span class="fu">write.gct</span>(</span>
<span id="cb5-17"><a href="#cb5-17"></a>    gex, </span>
<span id="cb5-18"><a href="#cb5-18"></a>    <span class="at">outdir =</span> <span class="st">"../1_processed/TCGA_lung_gex.gct"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-7" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>gs <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-2"><a href="#cb6-2"></a>gs[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../2_metastatic_genes/TCGA_lung_univariate_lasso.txt"</span>, <span class="at">header =</span> F)[[<span class="dv">1</span>]]</span>
<span id="cb6-3"><a href="#cb6-3"></a>gs[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../2_metastatic_genes/TCGA_LUSC_univariate_lasso.txt"</span>, <span class="at">header =</span> F)[[<span class="dv">1</span>]]</span>
<span id="cb6-4"><a href="#cb6-4"></a>gs[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../2_metastatic_genes/TCGA_LUAD_univariate_lasso.txt"</span>, <span class="at">header =</span> F)[[<span class="dv">1</span>]]</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="fu">names</span>(gs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"lung"</span>, <span class="st">"LUSC"</span>, <span class="st">"LUAD"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>write.gmt <span class="ot">&lt;-</span> <span class="cf">function</span>(genelist, file) {</span>
<span id="cb7-2"><a href="#cb7-2"></a>  file_conn <span class="ot">&lt;-</span> <span class="fu">file</span>(file, <span class="at">open =</span> <span class="st">"wt"</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a>  <span class="cf">for</span> (set_name <span class="cf">in</span> <span class="fu">names</span>(genelist)) {</span>
<span id="cb7-4"><a href="#cb7-4"></a>    line <span class="ot">&lt;-</span> <span class="fu">c</span>(set_name, set_name, genelist[[set_name]])</span>
<span id="cb7-5"><a href="#cb7-5"></a>    <span class="fu">writeLines</span>(<span class="fu">paste</span>(line, <span class="at">collapse =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>), file_conn)}</span>
<span id="cb7-6"><a href="#cb7-6"></a>  <span class="fu">close</span>(file_conn)}</span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="fu">write.gmt</span>(gs, <span class="st">"../2_metastatic_genes/metastatic_gene_list.gmt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="perform-ssgsea" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="perform-ssgsea"><span class="header-section-number">3.3</span> Perform ssGSEA</h2>
<div id="cell-10" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">library</span>(ssGSEA2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-11" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>res <span class="ot">=</span> <span class="fu">run_ssGSEA2</span>(</span>
<span id="cb9-2"><a href="#cb9-2"></a>    <span class="st">"../1_processed/TCGA_lung_gex.gct"</span>,</span>
<span id="cb9-3"><a href="#cb9-3"></a>    <span class="at">output.prefix =</span> <span class="st">"ssGSEA"</span>,</span>
<span id="cb9-4"><a href="#cb9-4"></a>    <span class="at">gene.set.databases =</span> <span class="st">"../2_metastatic_genes/metastatic_gene_list.gmt"</span>,</span>
<span id="cb9-5"><a href="#cb9-5"></a>    <span class="at">output.directory =</span> <span class="st">"../3_classification"</span>,</span>
<span id="cb9-6"><a href="#cb9-6"></a>    <span class="at">sample.norm.type =</span> <span class="st">"none"</span>, </span>
<span id="cb9-7"><a href="#cb9-7"></a>    <span class="at">weight =</span> <span class="fl">0.75</span>, </span>
<span id="cb9-8"><a href="#cb9-8"></a>    <span class="at">correl.type =</span> <span class="st">"rank"</span>, </span>
<span id="cb9-9"><a href="#cb9-9"></a>    <span class="at">statistic =</span> <span class="st">"area.under.RES"</span>,</span>
<span id="cb9-10"><a href="#cb9-10"></a>    <span class="at">output.score.type =</span> <span class="st">"NES"</span>, </span>
<span id="cb9-11"><a href="#cb9-11"></a>    <span class="at">nperm =</span> <span class="dv">1000</span>, </span>
<span id="cb9-12"><a href="#cb9-12"></a>    <span class="at">min.overlap =</span> <span class="dv">2</span>, </span>
<span id="cb9-13"><a href="#cb9-13"></a>    <span class="at">extended.output =</span> F, </span>
<span id="cb9-14"><a href="#cb9-14"></a>    <span class="at">global.fdr =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-15"><a href="#cb9-15"></a>    <span class="at">log.file =</span> <span class="st">"../3_classification/ssGSEA.log"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>parsing as GCT v1.2

../1_processed/TCGA_lung_gex.gct 20501 rows, 1017 cols, 0 row descriptors, 0 col descriptors
</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>lung   33.33333 %
LUSC   66.66667 %
LUAD   100 %
main loop: 1.832375 
Saving file to  ../3_classification/signature_gct/lung_n1017x151.gct 
Dimensions of matrix: [151x1017]
Setting precision to 4
Saved.
Saving file to  ../3_classification/signature_gct/LUSC_n1017x97.gct 
Dimensions of matrix: [97x1017]
Setting precision to 4
Saved.
Saving file to  ../3_classification/signature_gct/LUAD_n1017x110.gct 
Dimensions of matrix: [110x1017]
Setting precision to 4
Saved.
[1] "Total gene sets: 3"
[1] "Unique gene sets: 3"
Saving file to  ../3_classification/ssGSEA-scores.gct 
Dimensions of matrix: [3x1017]
Setting precision to 4
Saved.
Saving file to  ../3_classification/ssGSEA-pvalues.gct 
Dimensions of matrix: [3x1017]
Setting precision to 4
Saved.
Saving file to  ../3_classification/ssGSEA-fdr-pvalues.gct 
Dimensions of matrix: [3x1017]
Setting precision to 4
Saved.
Saving file to  ../3_classification/ssGSEA-combined.gct 
Dimensions of matrix: [3x1017]
Setting precision to 4
Saved.</code></pre>
</div>
</div>
</section>
<section id="combine-ssgsea-with-survival" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="combine-ssgsea-with-survival"><span class="header-section-number">3.4</span> Combine ssGSEA with survival</h2>
<div id="cell-13" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="37">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>ssgsea <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../3_classification/ssGSEA-scores.gct"</span>, <span class="at">row.names =</span> <span class="dv">1</span>, <span class="at">skip =</span> <span class="dv">2</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)]</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">colnames</span>(ssgsea) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="st">"-"</span>, <span class="fu">colnames</span>(ssgsea))</span>
<span id="cb12-3"><a href="#cb12-3"></a>ssgsea <span class="ot">&lt;-</span> <span class="fu">t</span>(ssgsea)</span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="fu">head</span>(ssgsea)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A matrix: 6 × 3 of type dbl</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">lung</th>
<th data-quarto-table-cell-role="th" scope="col">LUSC</th>
<th data-quarto-table-cell-role="th" scope="col">LUAD</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4244</td>
<td>8.5002</td>
<td>6.2464</td>
<td>5.3847</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4249</td>
<td>7.4015</td>
<td>6.3764</td>
<td>5.0068</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4250</td>
<td>8.0044</td>
<td>6.8932</td>
<td>5.9218</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4382</td>
<td>7.6315</td>
<td>6.2107</td>
<td>5.2858</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4384</td>
<td>8.0361</td>
<td>6.3106</td>
<td>6.1525</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4389</td>
<td>7.8553</td>
<td>6.4251</td>
<td>5.6181</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-14" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="38">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># read survival data</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>surv <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../1_processed/TCGA_lung_survival.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="fu">colnames</span>(surv) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="st">"_"</span>, <span class="fu">colnames</span>(surv))</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co"># remove time &lt; 0 or NA</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>surv <span class="ot">&lt;-</span> surv <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(PFI_time_1))</span>
<span id="cb13-7"><a href="#cb13-7"></a>surv<span class="sc">$</span>PFI_time_1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(surv<span class="sc">$</span>PFI_time_1)<span class="sc">/</span><span class="fl">365.25</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>surv<span class="sc">$</span>PFI_1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(surv<span class="sc">$</span>PFI_1)</span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a><span class="co"># set max PFI to 10 years</span></span>
<span id="cb13-11"><a href="#cb13-11"></a>surv<span class="sc">$</span>PFI_1 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(surv<span class="sc">$</span>PFI_time_1 <span class="sc">&gt;</span> <span class="dv">10</span>, <span class="dv">0</span>, surv<span class="sc">$</span>PFI_1)</span>
<span id="cb13-12"><a href="#cb13-12"></a>surv<span class="sc">$</span>PFI_time_1 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(surv<span class="sc">$</span>PFI_time_1 <span class="sc">&gt;</span> <span class="dv">10</span>, <span class="dv">10</span>, surv<span class="sc">$</span>PFI_time_1)</span>
<span id="cb13-13"><a href="#cb13-13"></a>surv <span class="ot">&lt;-</span> surv[surv<span class="sc">$</span>PFI_time_1 <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb13-14"><a href="#cb13-14"></a></span>
<span id="cb13-15"><a href="#cb13-15"></a><span class="co"># create survival object</span></span>
<span id="cb13-16"><a href="#cb13-16"></a>surv<span class="sc">$</span>PFI <span class="ot">&lt;-</span> <span class="fu">Surv</span>(surv<span class="sc">$</span>PFI_time_1, surv<span class="sc">$</span>PFI_1)</span>
<span id="cb13-17"><a href="#cb13-17"></a>surv <span class="ot">&lt;-</span> <span class="fu">cbind</span>(surv, ssgsea[<span class="fu">rownames</span>(surv),,<span class="at">drop =</span> F])</span>
<span id="cb13-18"><a href="#cb13-18"></a><span class="fu">head</span>(surv)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 21</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">type</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_1</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_1</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_2</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_2</th>
<th data-quarto-table-cell-role="th" scope="col">PFS</th>
<th data-quarto-table-cell-role="th" scope="col">PFS_time</th>
<th data-quarto-table-cell-role="th" scope="col">DSS_cr</th>
<th data-quarto-table-cell-role="th" scope="col">DSS_time_cr</th>
<th data-quarto-table-cell-role="th" scope="col">DFI_cr</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_1_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_1_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_2_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_2_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI</th>
<th data-quarto-table-cell-role="th" scope="col">lung</th>
<th data-quarto-table-cell-role="th" scope="col">LUSC</th>
<th data-quarto-table-cell-role="th" scope="col">LUAD</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;Surv&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4249</td>
<td>LUAD</td>
<td>0</td>
<td>4.1697467</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>NA</td>
<td>⋯</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>4.1697467, 0</td>
<td>7.4015</td>
<td>6.3764</td>
<td>5.0068</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4250</td>
<td>LUAD</td>
<td>0</td>
<td>0.3312799</td>
<td>0</td>
<td>121</td>
<td>1</td>
<td>121</td>
<td>NA</td>
<td>121</td>
<td>NA</td>
<td>⋯</td>
<td>2</td>
<td>121</td>
<td>2</td>
<td>121</td>
<td>2</td>
<td>121</td>
<td>0.3312799, 0</td>
<td>8.0044</td>
<td>6.8932</td>
<td>5.9218</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4382</td>
<td>LUAD</td>
<td>1</td>
<td>0.9144422</td>
<td>1</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>0</td>
<td>607</td>
<td>1</td>
<td>⋯</td>
<td>1</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>0.9144422, 1</td>
<td>7.6315</td>
<td>6.2107</td>
<td>5.2858</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4384</td>
<td>LUAD</td>
<td>1</td>
<td>0.5010267</td>
<td>1</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>0</td>
<td>426</td>
<td>NA</td>
<td>⋯</td>
<td>1</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>0.5010267, 1</td>
<td>8.0361</td>
<td>6.3106</td>
<td>6.1525</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4389</td>
<td>LUAD</td>
<td>0</td>
<td>3.7481177</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>NA</td>
<td>⋯</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>3.7481177, 0</td>
<td>7.8553</td>
<td>6.4251</td>
<td>5.6181</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4390</td>
<td>LUAD</td>
<td>1</td>
<td>1.0814511</td>
<td>NA</td>
<td>NA</td>
<td>1</td>
<td>395</td>
<td>0</td>
<td>1126</td>
<td>NA</td>
<td>⋯</td>
<td>1</td>
<td>395</td>
<td>1</td>
<td>395</td>
<td>NA</td>
<td>NA</td>
<td>1.0814511, 1</td>
<td>8.8213</td>
<td>6.5057</td>
<td>5.7669</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="lusc" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="lusc"><span class="header-section-number">3.5</span> LUSC</h2>
<p>Filter only LUSC patients and repeat with LUSC signatures.</p>
<div id="cell-16" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="39">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>find_coxph_cutoff <span class="ot">&lt;-</span> <span class="cf">function</span>(df, column, survival, <span class="at">set_cutoff =</span> <span class="cn">NULL</span>, <span class="at">covariates =</span> <span class="cn">NULL</span>, <span class="at">interquartile =</span> T){</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a>    <span class="co">#split data at 200 intervals and find the cutoff with most significant difference in survival</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>    <span class="cf">if</span>(<span class="fu">length</span>(covariates) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb14-5"><a href="#cb14-5"></a>        covariates_coxph <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">" + "</span>, <span class="fu">paste0</span>(covariates, <span class="at">collapse =</span> <span class="st">" + "</span>))</span>
<span id="cb14-6"><a href="#cb14-6"></a>        <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"set covariates ===&gt; "</span>, covariates_coxph))}</span>
<span id="cb14-7"><a href="#cb14-7"></a>    <span class="cf">else</span>{</span>
<span id="cb14-8"><a href="#cb14-8"></a>        covariates_coxph <span class="ot">&lt;-</span> <span class="st">""</span>}</span>
<span id="cb14-9"><a href="#cb14-9"></a>    cutoff <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-10"><a href="#cb14-10"></a></span>
<span id="cb14-11"><a href="#cb14-11"></a>    <span class="cf">if</span>(interquartile){</span>
<span id="cb14-12"><a href="#cb14-12"></a>        lower <span class="ot">&lt;-</span> <span class="fu">quantile</span>(df[[column]])[[<span class="dv">2</span>]]</span>
<span id="cb14-13"><a href="#cb14-13"></a>        upper <span class="ot">&lt;-</span> <span class="fu">quantile</span>(df[[column]])[[<span class="dv">4</span>]]}</span>
<span id="cb14-14"><a href="#cb14-14"></a>    <span class="cf">else</span>{</span>
<span id="cb14-15"><a href="#cb14-15"></a>        lower <span class="ot">&lt;-</span> <span class="fu">min</span>(df[[column]])</span>
<span id="cb14-16"><a href="#cb14-16"></a>    upper <span class="ot">&lt;-</span> <span class="fu">max</span>(df[[column]])}</span>
<span id="cb14-17"><a href="#cb14-17"></a></span>
<span id="cb14-18"><a href="#cb14-18"></a>    no <span class="ot">&lt;-</span> <span class="fu">nrow</span>(df)</span>
<span id="cb14-19"><a href="#cb14-19"></a></span>
<span id="cb14-20"><a href="#cb14-20"></a>    <span class="cf">for</span> (split <span class="cf">in</span> <span class="fu">seq</span>(lower, upper, <span class="at">length.out =</span> <span class="dv">200</span>)) {</span>
<span id="cb14-21"><a href="#cb14-21"></a>        df[[<span class="st">"coxph"</span>]] <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df[[column]] <span class="sc">&gt;</span> split, <span class="st">"High"</span>, <span class="st">"Low"</span>)</span>
<span id="cb14-22"><a href="#cb14-22"></a>        df[[<span class="st">"coxph"</span>]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df[[<span class="st">"coxph"</span>]], <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"High"</span>))</span>
<span id="cb14-23"><a href="#cb14-23"></a>        remove <span class="ot">&lt;-</span> <span class="fu">any</span>(<span class="fu">table</span>(df<span class="sc">$</span>coxph) <span class="sc">&lt;</span> no <span class="sc">/</span> <span class="dv">20</span>)</span>
<span id="cb14-24"><a href="#cb14-24"></a>        <span class="cf">if</span> (remove) {</span>
<span id="cb14-25"><a href="#cb14-25"></a>            <span class="cf">next</span></span>
<span id="cb14-26"><a href="#cb14-26"></a>        }</span>
<span id="cb14-27"><a href="#cb14-27"></a>        surv.cox <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">as.formula</span>(<span class="fu">paste0</span>(survival, <span class="st">" ~ coxph"</span>, covariates_coxph)), <span class="at">data =</span> df)</span>
<span id="cb14-28"><a href="#cb14-28"></a>        </span>
<span id="cb14-29"><a href="#cb14-29"></a>        <span class="co"># Extract HR and p-value</span></span>
<span id="cb14-30"><a href="#cb14-30"></a>        hr <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">exp</span>(<span class="fu">coef</span>(surv.cox))[<span class="dv">1</span>], <span class="dv">3</span>)  <span class="co"># Hazard Ratio</span></span>
<span id="cb14-31"><a href="#cb14-31"></a>        logr <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">summary</span>(surv.cox)<span class="sc">$</span>coefficients[<span class="dv">1</span>, <span class="st">"Pr(&gt;|z|)"</span>], <span class="dv">3</span>)  <span class="co"># p-value</span></span>
<span id="cb14-32"><a href="#cb14-32"></a>        </span>
<span id="cb14-33"><a href="#cb14-33"></a>        cutoff <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cutoff, <span class="fu">c</span>(split, hr, logr))</span>
<span id="cb14-34"><a href="#cb14-34"></a>        <span class="fu">colnames</span>(cutoff) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"cutoff"</span>, <span class="st">"HR"</span>, <span class="st">"logr"</span>)</span>
<span id="cb14-35"><a href="#cb14-35"></a>    }</span>
<span id="cb14-36"><a href="#cb14-36"></a></span>
<span id="cb14-37"><a href="#cb14-37"></a></span>
<span id="cb14-38"><a href="#cb14-38"></a>    <span class="co"># use most significant cutoff to split data in "High" &amp; "Low" groups</span></span>
<span id="cb14-39"><a href="#cb14-39"></a>    cutoff <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(cutoff) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(logr)</span>
<span id="cb14-40"><a href="#cb14-40"></a>    <span class="cf">if</span>(<span class="fu">length</span>(set_cutoff) <span class="sc">==</span> <span class="dv">0</span>){</span>
<span id="cb14-41"><a href="#cb14-41"></a>        threshold <span class="ot">&lt;-</span> cutoff<span class="sc">$</span>cutoff[<span class="dv">1</span>]}</span>
<span id="cb14-42"><a href="#cb14-42"></a>    <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">length</span>(set_cutoff) <span class="sc">==</span> <span class="dv">1</span>){</span>
<span id="cb14-43"><a href="#cb14-43"></a>        threshold <span class="ot">&lt;-</span> set_cutoff}</span>
<span id="cb14-44"><a href="#cb14-44"></a>    <span class="cf">else</span>{</span>
<span id="cb14-45"><a href="#cb14-45"></a>    <span class="fu">error</span>(<span class="st">"set_cutoff should be set to length of 1"</span>)}</span>
<span id="cb14-46"><a href="#cb14-46"></a>    <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"for &lt;"</span>, column, <span class="st">"&gt;, cutoff is set at  ===&gt; "</span>, threshold))</span>
<span id="cb14-47"><a href="#cb14-47"></a>    df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]] <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df[[column]] <span class="sc">&gt;</span> <span class="fu">paste0</span>(threshold), <span class="st">"High"</span>, <span class="st">"Low"</span>)</span>
<span id="cb14-48"><a href="#cb14-48"></a>    df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]], <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"High"</span>))</span>
<span id="cb14-49"><a href="#cb14-49"></a>    df<span class="sc">$</span>coxph <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-50"><a href="#cb14-50"></a>    <span class="fu">return</span>(df)</span>
<span id="cb14-51"><a href="#cb14-51"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-17" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="40">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>surv.LUSC <span class="ot">&lt;-</span> surv <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2"></a>    <span class="fu">filter</span>(type <span class="sc">==</span> <span class="st">"LUSC"</span>)</span>
<span id="cb15-3"><a href="#cb15-3"></a>surv.LUSC <span class="ot">&lt;-</span> <span class="fu">find_coxph_cutoff</span>(<span class="at">df =</span> surv.LUSC, <span class="at">column =</span> <span class="st">"LUSC"</span>, <span class="at">survival =</span> <span class="st">"PFI"</span>, <span class="at">covariate =</span> <span class="cn">NULL</span>, <span class="at">interquartile =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "for &lt;LUSC&gt;, cutoff is set at  ===&gt; 6.36844773869347"</code></pre>
</div>
</div>
<div id="cell-18" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="41">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_survival</span>(surv.LUSC, <span class="at">column =</span> <span class="st">"LUSC_coxph"</span>, <span class="at">legend.title =</span> <span class="st">"LUSC"</span>, <span class="at">survival =</span> <span class="st">"PFI"</span>, <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"grey60"</span>, <span class="st">"grey1"</span>), <span class="at">conf.int =</span> T, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">ylab =</span> <span class="st">"Probability"</span>, <span class="at">xlab =</span> <span class="st">"PFI (Years)"</span>, <span class="at">pval.coord =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.1</span>))</span>
<span id="cb17-2"><a href="#cb17-2"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“The `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.
ℹ Please use the `linewidth` argument instead.”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="ssgsea2_files/figure-html/cell-14-output-2.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-19" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="42">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>surv <span class="ot">&lt;-</span> <span class="fu">merge</span>(surv, surv.LUSC[, <span class="fu">c</span>(<span class="st">"LUSC_coxph"</span>), <span class="at">drop =</span> F], <span class="at">by =</span> <span class="dv">0</span>, <span class="at">all.x =</span> T) <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"Row.names"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="luad" class="level2" data-number="3.6">
<h2 data-number="3.6" class="anchored" data-anchor-id="luad"><span class="header-section-number">3.6</span> LUAD</h2>
<p>Filter only LUAD patients and repeat with LUSC signatures.</p>
<div id="cell-21" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="43">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>find_coxph_cutoff <span class="ot">&lt;-</span> <span class="cf">function</span>(df, column, survival, <span class="at">set_cutoff =</span> <span class="cn">NULL</span>, <span class="at">covariates =</span> <span class="cn">NULL</span>, <span class="at">interquartile =</span> T){</span>
<span id="cb20-2"><a href="#cb20-2"></a></span>
<span id="cb20-3"><a href="#cb20-3"></a>    <span class="co">#split data at 200 intervals and find the cutoff with most significant difference in survival</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>    <span class="cf">if</span>(<span class="fu">length</span>(covariates) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb20-5"><a href="#cb20-5"></a>        covariates_coxph <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">" + "</span>, <span class="fu">paste0</span>(covariates, <span class="at">collapse =</span> <span class="st">" + "</span>))</span>
<span id="cb20-6"><a href="#cb20-6"></a>        <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"set covariates ===&gt; "</span>, covariates_coxph))}</span>
<span id="cb20-7"><a href="#cb20-7"></a>    <span class="cf">else</span>{</span>
<span id="cb20-8"><a href="#cb20-8"></a>        covariates_coxph <span class="ot">&lt;-</span> <span class="st">""</span>}</span>
<span id="cb20-9"><a href="#cb20-9"></a>    cutoff <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb20-10"><a href="#cb20-10"></a></span>
<span id="cb20-11"><a href="#cb20-11"></a>    <span class="cf">if</span>(interquartile){</span>
<span id="cb20-12"><a href="#cb20-12"></a>        lower <span class="ot">&lt;-</span> <span class="fu">quantile</span>(df[[column]])[[<span class="dv">2</span>]]</span>
<span id="cb20-13"><a href="#cb20-13"></a>        upper <span class="ot">&lt;-</span> <span class="fu">quantile</span>(df[[column]])[[<span class="dv">4</span>]]}</span>
<span id="cb20-14"><a href="#cb20-14"></a>    <span class="cf">else</span>{</span>
<span id="cb20-15"><a href="#cb20-15"></a>        lower <span class="ot">&lt;-</span> <span class="fu">min</span>(df[[column]])</span>
<span id="cb20-16"><a href="#cb20-16"></a>    upper <span class="ot">&lt;-</span> <span class="fu">max</span>(df[[column]])}</span>
<span id="cb20-17"><a href="#cb20-17"></a></span>
<span id="cb20-18"><a href="#cb20-18"></a>    no <span class="ot">&lt;-</span> <span class="fu">nrow</span>(df)</span>
<span id="cb20-19"><a href="#cb20-19"></a></span>
<span id="cb20-20"><a href="#cb20-20"></a>    <span class="cf">for</span> (split <span class="cf">in</span> <span class="fu">seq</span>(lower, upper, <span class="at">length.out =</span> <span class="dv">200</span>)) {</span>
<span id="cb20-21"><a href="#cb20-21"></a>        df[[<span class="st">"coxph"</span>]] <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df[[column]] <span class="sc">&gt;</span> split, <span class="st">"High"</span>, <span class="st">"Low"</span>)</span>
<span id="cb20-22"><a href="#cb20-22"></a>        df[[<span class="st">"coxph"</span>]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df[[<span class="st">"coxph"</span>]], <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"High"</span>))</span>
<span id="cb20-23"><a href="#cb20-23"></a>        remove <span class="ot">&lt;-</span> <span class="fu">any</span>(<span class="fu">table</span>(df<span class="sc">$</span>coxph) <span class="sc">&lt;</span> no <span class="sc">/</span> <span class="dv">20</span>)</span>
<span id="cb20-24"><a href="#cb20-24"></a>        <span class="cf">if</span> (remove) {</span>
<span id="cb20-25"><a href="#cb20-25"></a>            <span class="cf">next</span></span>
<span id="cb20-26"><a href="#cb20-26"></a>        }</span>
<span id="cb20-27"><a href="#cb20-27"></a>        surv.cox <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">as.formula</span>(<span class="fu">paste0</span>(survival, <span class="st">" ~ coxph"</span>, covariates_coxph)), <span class="at">data =</span> df)</span>
<span id="cb20-28"><a href="#cb20-28"></a>        </span>
<span id="cb20-29"><a href="#cb20-29"></a>        <span class="co"># Extract HR and p-value</span></span>
<span id="cb20-30"><a href="#cb20-30"></a>        hr <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">exp</span>(<span class="fu">coef</span>(surv.cox))[<span class="dv">1</span>], <span class="dv">3</span>)  <span class="co"># Hazard Ratio</span></span>
<span id="cb20-31"><a href="#cb20-31"></a>        logr <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">summary</span>(surv.cox)<span class="sc">$</span>coefficients[<span class="dv">1</span>, <span class="st">"Pr(&gt;|z|)"</span>], <span class="dv">3</span>)  <span class="co"># p-value</span></span>
<span id="cb20-32"><a href="#cb20-32"></a>        </span>
<span id="cb20-33"><a href="#cb20-33"></a>        cutoff <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cutoff, <span class="fu">c</span>(split, hr, logr))</span>
<span id="cb20-34"><a href="#cb20-34"></a>        <span class="fu">colnames</span>(cutoff) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"cutoff"</span>, <span class="st">"HR"</span>, <span class="st">"logr"</span>)</span>
<span id="cb20-35"><a href="#cb20-35"></a>    }</span>
<span id="cb20-36"><a href="#cb20-36"></a></span>
<span id="cb20-37"><a href="#cb20-37"></a></span>
<span id="cb20-38"><a href="#cb20-38"></a>    <span class="co"># use most significant cutoff to split data in "High" &amp; "Low" groups</span></span>
<span id="cb20-39"><a href="#cb20-39"></a>    cutoff <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(cutoff) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(logr)</span>
<span id="cb20-40"><a href="#cb20-40"></a>    <span class="cf">if</span>(<span class="fu">length</span>(set_cutoff) <span class="sc">==</span> <span class="dv">0</span>){</span>
<span id="cb20-41"><a href="#cb20-41"></a>        threshold <span class="ot">&lt;-</span> cutoff<span class="sc">$</span>cutoff[<span class="dv">1</span>]}</span>
<span id="cb20-42"><a href="#cb20-42"></a>    <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">length</span>(set_cutoff) <span class="sc">==</span> <span class="dv">1</span>){</span>
<span id="cb20-43"><a href="#cb20-43"></a>        threshold <span class="ot">&lt;-</span> set_cutoff}</span>
<span id="cb20-44"><a href="#cb20-44"></a>    <span class="cf">else</span>{</span>
<span id="cb20-45"><a href="#cb20-45"></a>    <span class="fu">error</span>(<span class="st">"set_cutoff should be set to length of 1"</span>)}</span>
<span id="cb20-46"><a href="#cb20-46"></a>    <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"for &lt;"</span>, column, <span class="st">"&gt;, cutoff is set at  ===&gt; "</span>, threshold))</span>
<span id="cb20-47"><a href="#cb20-47"></a>    df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]] <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df[[column]] <span class="sc">&gt;</span> <span class="fu">paste0</span>(threshold), <span class="st">"High"</span>, <span class="st">"Low"</span>)</span>
<span id="cb20-48"><a href="#cb20-48"></a>    df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df[[<span class="fu">paste0</span>(column, <span class="st">"_coxph"</span>)]], <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"High"</span>))</span>
<span id="cb20-49"><a href="#cb20-49"></a>    df<span class="sc">$</span>coxph <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb20-50"><a href="#cb20-50"></a>    <span class="fu">return</span>(df)</span>
<span id="cb20-51"><a href="#cb20-51"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-22" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="44">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>surv.LUAD <span class="ot">&lt;-</span> surv <span class="sc">%&gt;%</span> </span>
<span id="cb21-2"><a href="#cb21-2"></a>    <span class="fu">filter</span>(type <span class="sc">==</span> <span class="st">"LUAD"</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>surv.LUAD <span class="ot">&lt;-</span> <span class="fu">find_coxph_cutoff</span>(<span class="at">df =</span> surv.LUAD, <span class="at">column =</span> <span class="st">"LUAD"</span>, <span class="at">survival =</span> <span class="st">"PFI"</span>, <span class="at">covariate =</span> <span class="cn">NULL</span>, <span class="at">interquartile =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "for &lt;LUAD&gt;, cutoff is set at  ===&gt; 5.2353"</code></pre>
</div>
</div>
<div id="cell-23" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="45">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">head</span>(surv)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 22</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">type</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_1</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_1</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_2</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_2</th>
<th data-quarto-table-cell-role="th" scope="col">PFS</th>
<th data-quarto-table-cell-role="th" scope="col">PFS_time</th>
<th data-quarto-table-cell-role="th" scope="col">DSS_cr</th>
<th data-quarto-table-cell-role="th" scope="col">DSS_time_cr</th>
<th data-quarto-table-cell-role="th" scope="col">DFI_cr</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_1_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_1_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_2_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI_time_2_cr</th>
<th data-quarto-table-cell-role="th" scope="col">PFI</th>
<th data-quarto-table-cell-role="th" scope="col">lung</th>
<th data-quarto-table-cell-role="th" scope="col">LUSC</th>
<th data-quarto-table-cell-role="th" scope="col">LUAD</th>
<th data-quarto-table-cell-role="th" scope="col">LUSC_coxph</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;Surv&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4249</td>
<td>LUAD</td>
<td>0</td>
<td>4.1697467</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>NA</td>
<td>⋯</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>0</td>
<td>1523</td>
<td>4.1697467, 0</td>
<td>7.4015</td>
<td>6.3764</td>
<td>5.0068</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4250</td>
<td>LUAD</td>
<td>0</td>
<td>0.3312799</td>
<td>0</td>
<td>121</td>
<td>1</td>
<td>121</td>
<td>NA</td>
<td>121</td>
<td>NA</td>
<td>⋯</td>
<td>121</td>
<td>2</td>
<td>121</td>
<td>2</td>
<td>121</td>
<td>0.3312799, 0</td>
<td>8.0044</td>
<td>6.8932</td>
<td>5.9218</td>
<td>NA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4382</td>
<td>LUAD</td>
<td>1</td>
<td>0.9144422</td>
<td>1</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>0</td>
<td>607</td>
<td>1</td>
<td>⋯</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>1</td>
<td>334</td>
<td>0.9144422, 1</td>
<td>7.6315</td>
<td>6.2107</td>
<td>5.2858</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4384</td>
<td>LUAD</td>
<td>1</td>
<td>0.5010267</td>
<td>1</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>0</td>
<td>426</td>
<td>NA</td>
<td>⋯</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>1</td>
<td>183</td>
<td>0.5010267, 1</td>
<td>8.0361</td>
<td>6.3106</td>
<td>6.1525</td>
<td>NA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4389</td>
<td>LUAD</td>
<td>0</td>
<td>3.7481177</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>NA</td>
<td>⋯</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>0</td>
<td>1369</td>
<td>3.7481177, 0</td>
<td>7.8553</td>
<td>6.4251</td>
<td>5.6181</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4390</td>
<td>LUAD</td>
<td>1</td>
<td>1.0814511</td>
<td>NA</td>
<td>NA</td>
<td>1</td>
<td>395</td>
<td>0</td>
<td>1126</td>
<td>NA</td>
<td>⋯</td>
<td>395</td>
<td>1</td>
<td>395</td>
<td>NA</td>
<td>NA</td>
<td>1.0814511, 1</td>
<td>8.8213</td>
<td>6.5057</td>
<td>5.7669</td>
<td>NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-24" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="46">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>plot_survival <span class="ot">&lt;-</span> <span class="cf">function</span>(df, column, survival, <span class="at">palette =</span> <span class="cn">NULL</span>, ...){</span>
<span id="cb24-2"><a href="#cb24-2"></a>    <span class="cf">if</span>(<span class="sc">!</span>(column <span class="sc">%in%</span> <span class="fu">colnames</span>(df))){</span>
<span id="cb24-3"><a href="#cb24-3"></a>    <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"&lt;"</span>, column, <span class="st">"&gt; not found in dataframe. Skipped"</span>))</span>
<span id="cb24-4"><a href="#cb24-4"></a>    <span class="cf">next</span>}</span>
<span id="cb24-5"><a href="#cb24-5"></a></span>
<span id="cb24-6"><a href="#cb24-6"></a>    <span class="fu">suppressMessages</span>({</span>
<span id="cb24-7"><a href="#cb24-7"></a>    b <span class="ot">&lt;-</span> <span class="fu">as.formula</span>(<span class="fu">paste0</span>(survival, <span class="st">"~"</span>, column))</span>
<span id="cb24-8"><a href="#cb24-8"></a>    fit <span class="ot">&lt;-</span> <span class="fu">eval</span>(<span class="fu">substitute</span>(<span class="fu">survfit</span>(b, <span class="at">data =</span> df, <span class="at">conf.type =</span> <span class="st">"log-log"</span>), <span class="fu">list</span>(<span class="at">b =</span> b)))</span>
<span id="cb24-9"><a href="#cb24-9"></a></span>
<span id="cb24-10"><a href="#cb24-10"></a>    <span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">names</span>(fit<span class="sc">$</span>strata)) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb24-11"><a href="#cb24-11"></a>        <span class="fu">names</span>(fit<span class="sc">$</span>strata) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">".*="</span>, <span class="st">""</span>, <span class="fu">names</span>(fit<span class="sc">$</span>strata))</span>
<span id="cb24-12"><a href="#cb24-12"></a>        kmplot <span class="ot">&lt;-</span> <span class="fu">ggsurvplot</span>(</span>
<span id="cb24-13"><a href="#cb24-13"></a>            fit,</span>
<span id="cb24-14"><a href="#cb24-14"></a>            <span class="at">data =</span> df,</span>
<span id="cb24-15"><a href="#cb24-15"></a>            <span class="at">palette =</span> palette,</span>
<span id="cb24-16"><a href="#cb24-16"></a>            <span class="at">risk.table =</span> T,</span>
<span id="cb24-17"><a href="#cb24-17"></a>            <span class="at">pval =</span> T,</span>
<span id="cb24-18"><a href="#cb24-18"></a>            <span class="at">risk.table.y.text.col =</span> T,</span>
<span id="cb24-19"><a href="#cb24-19"></a>            <span class="at">risk.table.y.text =</span> F,</span>
<span id="cb24-20"><a href="#cb24-20"></a>            <span class="at">legend =</span> <span class="st">"right"</span>,</span>
<span id="cb24-21"><a href="#cb24-21"></a>            ...)</span>
<span id="cb24-22"><a href="#cb24-22"></a>            }</span>
<span id="cb24-23"><a href="#cb24-23"></a>    <span class="cf">else</span>{</span>
<span id="cb24-24"><a href="#cb24-24"></a>            kmplot <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb24-25"><a href="#cb24-25"></a>    }})</span>
<span id="cb24-26"><a href="#cb24-26"></a>    kmplot[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> kmplot[[<span class="dv">1</span>]] <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">size=</span><span class="fl">1.5</span>))</span>
<span id="cb24-27"><a href="#cb24-27"></a>    kmplot[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> kmplot[[<span class="dv">2</span>]] <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill=</span><span class="cn">NA</span>, <span class="at">size=</span><span class="fl">1.5</span>))</span>
<span id="cb24-28"><a href="#cb24-28"></a>    <span class="fu">return</span>(kmplot)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-25" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="47">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_survival</span>(surv.LUAD, <span class="at">column =</span> <span class="st">"LUAD_coxph"</span>, <span class="at">legend.title =</span> <span class="st">"LUAD"</span>, <span class="at">survival =</span> <span class="st">"PFI"</span>, <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"grey60"</span>, <span class="st">"grey1"</span>), <span class="at">conf.int =</span> T, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">ylab =</span> <span class="st">"Probability"</span>, <span class="at">xlab =</span> <span class="st">"PFI (Years)"</span>, <span class="at">pval.coord =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.1</span>))</span>
<span id="cb25-2"><a href="#cb25-2"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="ssgsea2_files/figure-html/cell-20-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-26" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="24">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>surv <span class="ot">&lt;-</span> <span class="fu">merge</span>(surv, surv.LUAD[, <span class="fu">c</span>(<span class="st">"LUAD_coxph"</span>), <span class="at">drop =</span> F], <span class="at">by =</span> <span class="dv">0</span>, <span class="at">all.x =</span> T) <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"Row.names"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="save-output" class="level2" data-number="3.7">
<h2 data-number="3.7" class="anchored" data-anchor-id="save-output"><span class="header-section-number">3.7</span> Save Output</h2>
<div id="cell-28" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="33">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>class <span class="ot">&lt;-</span> surv[,<span class="fu">c</span>(<span class="st">"LUSC_coxph"</span>, <span class="st">"LUAD_coxph"</span>), drop <span class="ot">=</span> F]</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="cf">for</span>(c <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(class)){</span>
<span id="cb27-3"><a href="#cb27-3"></a>    class[,c] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(class[,c])}</span>
<span id="cb27-4"><a href="#cb27-4"></a>class[class <span class="sc">==</span> <span class="st">"High"</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb27-5"><a href="#cb27-5"></a>class[class <span class="sc">==</span> <span class="st">"Low"</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="fu">head</span>(class)</span>
<span id="cb27-7"><a href="#cb27-7"></a></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="fu">write.csv</span>(class, <span class="st">"../3_classification/TCGA_metastatic_classification.csv"</span>, <span class="at">quote =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">LUSC_coxph</th>
<th data-quarto-table-cell-role="th" scope="col">LUAD_coxph</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4249</td>
<td>NA</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4250</td>
<td>NA</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4382</td>
<td>NA</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4384</td>
<td>NA</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4389</td>
<td>NA</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">TCGA-05-4390</td>
<td>NA</td>
<td>1</td>
</tr>
</tbody>
</table>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./metastatic_signature2.html" class="pagination-link" aria-label="Generating Prognostic Signature">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Generating Prognostic Signature</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./xenium2.html" class="pagination-link" aria-label="Validation in Xenium">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Validation in Xenium</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.dev/FrancisCrickInstitute/blob/main/ssgsea2.ipynb" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div></div></div></footer><script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>